src/slurm_plugin/fleet_manager.py

```diff
@@ -0,0 +1,300 @@
+        }
+
+        launch_params.update(launch_overrides)
+        if launch_overrides:
+            logger.info("Found CreateFleet parameters override. Launching instances with: %s", launch_params)
+        return launch_params
+
+    def _launch_instances(self, launch_params):
+        """Launch a batch of ec2 instances."""
+        try:
+            response = create_fleet(self._region, self._boto3_config, launch_params)
+            for error in response.get("Errors", []):
+                logger.error("Error in CreateFleet request: %s - %s", error.get("ErrorCode"), error.get("ErrorMessage"))
+
+            logger.debug("CreateFleet response: %s", response)
+            instance_ids = [inst_id for instances in response["Instances"] for inst_id in instances["InstanceIds"]]
+            return {"Instances": self._get_instances_info(instance_ids)}
+        except ClientError as e:
+            logger.error("Failed CreateFleet request: %s", e.response.get("ResponseMetadata").get("RequestId"))
+            raise e
+
+    def _get_instances_info(self, instance_ids, retry=3):
+        """
+        Describe instances to retrieve info not available from create-fleet response.
+
+        :raises EC2InstancesInfoUnavailable if after N retries it's not possible to retrieve info for all the instances
+        :raises ClientError in case of boto3 failure
+        """
+        # Wait for instances to be available in EC2
+        time.sleep(5)
+
+        # Retrieve instance info from Instance Ids
+        instances = []
+        if instance_ids:
+            logger.info(f"Retrieving instance info for instances: {instance_ids}")
+            ec2_client = boto3.client("ec2", region_name=self._region, config=self._boto3_config)
+            paginator = ec2_client.get_paginator("describe_instances")
+            response_iterator = paginator.paginate(InstanceIds=instance_ids)
+            filtered_iterator = response_iterator.search("Reservations[].Instances[]")
+            instances = [instance_info for instance_info in filtered_iterator]
+            logging.debug("Launched instances: %s", instances)
+
+            partial_info_instances = []
+            for instance_info in instances:
+                try:
+                    # Try to build EC2Instance objects retrieving all the required fields
+                    EC2Instance.from_info(instance_info)
+                except KeyError:
+                    failed_instance_id = instance_info["InstanceId"]
+                    partial_info_instances.append(failed_instance_id)
+                    instances.remove(instance_info)
+                    logging.warning(f"Unable to retrieve all the required fields for instance {failed_instance_id}")
+
+            if partial_info_instances:
+                # Try to find information for incomplete instances up to 3 times
+                if retry != 0:
+                    instances.extend(self._get_instances_info(partial_info_instances, retry - 1))
```

Comment:
rather than calling this function recursively why don't we separate the logic in two functions, one that tries to retrieve the info and another one that manages the retry? it should make the entire code more readable by separating the responsibilities

Reply:
Partially done. Let's talk about this.

Reply:
Done by creating a private recursive function.